### PR TITLE
Leniz_Khusaenov_Task_1_Middle+

### PR DESCRIPTION
**Pull request by LennyKhusaenov**
Ip counter was implemented:
- business logic was implemented, the solution is based on the limitation of ip v4
- tests were added
- code coverage was configured by jacoco
- pipeline based on github ci was included
- README.md was written

### CODE REVIEW

File: src/test/java/com/codereview/ipcounter/utils/ParserUtilsUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import org.junit.jupiter.params.provider.CsvFileSource;
+
+class ParserUtilsUnitTest {
+
+  @SneakyThrows
+  @ParameterizedTest
+  @CsvFileSource(resources = "/ip-to-long-view.csv", numLinesToSkip = 1)
+  void shouldTransformIpToLong(String testIp, String expected) {
+    //when
+    var result = ParserUtils.toLongValue(testIp);
```
Comment (MadPigeon) [inline src/test/java/com/codereview/ipcounter/utils/ParserUtilsUnitTest.java:17]: Unhandled exception type UnknownHostException.
Please surround the method with `try{} catch(){}` with fails in case it's caught.

File: src/test/java/com/codereview/ipcounter/utils/ParserUtilsUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package com.codereview.ipcounter.utils;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+import lombok.SneakyThrows;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvFileSource;
+
+class ParserUtilsUnitTest {
```
Comment (MadPigeon) [inline src/test/java/com/codereview/ipcounter/utils/ParserUtilsUnitTest.java:10]: Check edge cases:

- [ ] faulty IPs like `500.500.500.500`
- [ ] cases like `localhost`
- [ ] cases where ip has port like `127.4.5.250:8080`
- [ ] non-ips like `error-line`
- [ ] short ips like `127.0.0`

All of those cases are nice to have in test class so that people know how to use them and what to expect

File: src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  private FileIpDataProcessor fileIpDataProcessor;
+
+  @TempDir
+  private Path tempDir;
+
+  @SneakyThrows
+  @Test
+  void shouldProcessFile() {
+    //given
+    var tempFilePath = Files.createFile(tempDir.resolve(TEST_TXT));
```
Comment (MadPigeon) [inline src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java:43]: Unhandled exception type IOException
Surround the call with `try{}catch(){}` and assert fail the caught version

File: src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    //then
+    verify(ipCounter, times(3)).countIp(anyString());
+  }
+
+  @SneakyThrows
+  @Test
+  void shouldStopProcessIfAllIpsWereRegistered() {
+    //given
+    var tempFilePath = Files.createFile(tempDir.resolve(TEST_TXT));
```
Comment (MadPigeon) [inline src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java:61]: Unhandled exception type IOException
Surround the call with `try{}catch(){}` and assert fail the caught version

File: src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    //then
+    verify(ipCounter, times(0)).countIp(anyString());
+  }
+
+  @SneakyThrows
+  @Test
+  void shouldStopProcessIfEofReceived() {
+    //given
+    var tempFilePath = Files.createFile(tempDir.resolve(TEST_TXT));
```
Comment (MadPigeon) [inline src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java:79]: Unhandled exception type IOException
Surround the call with `try{}catch(){}` and assert fail the caught version

File: src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    //when
+    fileIpDataProcessor.processFile(tempFilePath.toString());
+
+    //then
+    verify(ipCounter, times(3)).countIp(anyString());
+  }
+
+  @SneakyThrows
+  @Test
+  void shouldStopProcessIfAllIpsWereRegistered() {
```
Comment (MadPigeon) [inline src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java:59]: This test is confusing. Why shouldn't `countIp` method be called at all?

File: src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+import lombok.SneakyThrows;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.io.TempDir;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class FileIpDataProcessorUnitTest {
```
Comment (MadPigeon) [inline src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java:28]: After reading that class I still don't know how your program will react to a file with bad values

File: src/test/java/com/codereview/ipcounter/service/FileNameArgumentProcessorUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+    //when
+    fileNameArgumentProcessor.process(applicationArguments);
+
+    //then
+    verify(fileDataProcessor, times(1)).processFile(fileName.trim());
+  }
+
+  @Test
+  void shouldNotIfArgsAreEmpty() {
```
Comment (MadPigeon) [inline src/test/java/com/codereview/ipcounter/service/FileNameArgumentProcessorUnitTest.java:74]: should not... **what?** if args are empty?

File: src/test/java/com/codereview/ipcounter/service/UniqueIpCounterUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    ipCounter.resetCounter();
+  }
+
+  static Stream<Arguments> createCountToIpsParams() {
+    return Stream.of(
+        arguments(3L, List.of("127.0.0.1",
+            "223.223.223.223",
+            "255.255.255.255",
+            "255.255.255.255")),
+        arguments(1L, List.of("")),
```
Comment (MadPigeon) [inline src/test/java/com/codereview/ipcounter/service/UniqueIpCounterUnitTest.java:37]: Do you expect your program to output that you have found one unique IP when you are given an empty string?
I find that functionality weird.
Or I am misunderstanding and would like an explanation

File: README.md [hunk 10]
```diff
@@ -1 +1,42 @@
...
+2. Parse long view of ip bytes
+3. Calculate integer index of ip in array by dividing this view by 32
+4. If we have no value at this position, store it and increment count
+
+</p>
+
+### Application start
+
+Application start command:<br>
+`mvnw spring-boot:run -Dspring-boot.run.arguments=--file.path={YOUR_FILE_PATH_HERE}`
```
Comment (MadPigeon) [inline README.md:24]: Tried that, ended up getting an error despite being in the correct directory. Amazing that you have a well-written readme though.
>mvnw : The term 'mvnw' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is 
correct and try again.

File: src/main/java/com/codereview/ipcounter/utils/ParserUtils.java [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+public class ParserUtils {
+  public static long toLongValue(String ipString) throws UnknownHostException {
+    long result = 0;
+    for (byte b : InetAddress.getByName(ipString).getAddress()) {
+      result = (result << 8) | (b & 255);
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/utils/ParserUtils.java:12]: this looks efficient but hard to read and check validity

File: src/main/java/com/codereview/ipcounter/service/IpCounter.java [hunk 12]
```diff
@@ -0,0 +1,10 @@
+package com.codereview.ipcounter.service;
+
+public interface IpCounter<T> {
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/service/IpCounter.java:3]: You're using only `<String>`, why have `<T>` here?

File: src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+  private static final String TEST_TXT = "test.txt";
+  @Mock
+  private IpCounter<String> ipCounter;
+  @InjectMocks
+  private FileIpDataProcessor fileIpDataProcessor;
```
Comment (MadPigeon) [inline src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java:34]: Why are you using the `IPCounter` interface and are not using the `FileDataProcessor` interface?

File: src/main/java/com/codereview/ipcounter/service/ArgumentProcessor.java [hunk 14]
```diff
@@ -0,0 +1,9 @@
+package com.codereview.ipcounter.service;
+
+import org.springframework.boot.ApplicationArguments;
+
+@FunctionalInterface
+public interface ArgumentProcessor {
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/service/ArgumentProcessor.java:6]: I don't understand why do you need this interface

File: src/main/java/com/codereview/ipcounter/service/impl/FileNameArgumentProcessor.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+import com.codereview.ipcounter.service.FileDataProcessor;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.boot.ApplicationArguments;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class FileNameArgumentProcessor implements ArgumentProcessor {
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/service/impl/FileNameArgumentProcessor.java:16]: I don't understand why you need this class. It's only used in tests.

File: src/main/java/com/codereview/ipcounter/service/impl/FileNameArgumentProcessor.java [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+  private final FileDataProcessor fileDataProcessor;
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/service/impl/FileNameArgumentProcessor.java:18]: Warning
>The blank final field fileDataProcessor may not have been initialized

File: src/main/java/com/codereview/ipcounter/service/impl/FileNameArgumentProcessor.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+
+  @Override
+  public void process(ApplicationArguments applicationArguments) {
+    var optionValues = applicationArguments.getOptionValues(FILE_NAME_ARG);
+    if (isEmpty(optionValues)) {
+      log.error("Can't find file name argument");
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/service/impl/FileNameArgumentProcessor.java:24]: Warning
>log cannot be resolved

(same with line 27)

File: src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.context.event.EventListener;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class IpCounterEventHandler {
+
+  private final ApplicationArguments applicationArguments;
+  private final List<ArgumentProcessor> argumentProcessors;
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java:18]: Warning
>The blank final field argumentProcessors may not have been initialized

Why do you need an empty list that's not initialized in any constructor?

File: src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.boot.context.event.ApplicationReadyEvent;
+import org.springframework.context.event.EventListener;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class IpCounterEventHandler {
+
+  private final ApplicationArguments applicationArguments;
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java:17]: Warning
>The blank final field applicationArguments may not have been initialized

File: src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.boot.ApplicationArguments;
+import org.springframework.boot.context.event.ApplicationReadyEvent;
+import org.springframework.context.event.EventListener;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class IpCounterEventHandler {
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java:15]: I don't understand why you need this class. I would understand seeing it in tests.

File: src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.stream.Stream;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class FileIpDataProcessor implements FileDataProcessor {
+  private final IpCounter<String> ipCounter;
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java:20]: Warning
>The blank final field ipCounter may not have been initialized

File: src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+  @Override
+  public void processFile(String filePath) {
+    try (Stream<String> ipLines = Files.lines(Paths.get(filePath))) {
+      var linesProcessed = new AtomicInteger();
+      ipLines.takeWhile(line -> ipCounter.getCount() < NUMBER_OF_IP_ADDRESSES)
+          .forEach(ipView -> {
+            ipCounter.countIp(ipView);
+            linesProcessed.incrementAndGet();
+          });
+      log.info("Total lines processed: {} and unique ip count: {}", linesProcessed,
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java:31]: Warning
>log cannot be resolved

Same with line 34

File: src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+
+  private long count;
+
+  @Override
+  public void countIp(String ipView) {
+    try {
+      var longIpView = ParserUtils.toLongValue(ipView);
+      registerIfNewUnique(longIpView);
+    } catch (UnknownHostException e) {
+      log.error("An error occurred during try to parse ip {}", ipView, e);
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java:23]: Warning
>log cannot be resolved

File: src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+import com.codereview.ipcounter.utils.ParserUtils;
+import java.net.UnknownHostException;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+public class UniqueIpCounter implements IpCounter<String> {
+
+  private final int[] seen = new int[1 << 27];
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java:13]: This looks very smart. I don't understand what's happening here. Add a comment, please

File: src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+  public long getCount() {
+    return count;
+  }
+
+  @Override
+  public void resetCounter() {
+    count = 0;
+  }
+
+  private void registerIfNewUnique(long longValue) {
```
Comment (MadPigeon) [inline src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java:37]: those bit operations look very smart, so I am confused. I don't understand what the code does or why it does it

**Review by MadPigeon — CHANGES_REQUESTED** [review]
- [ ] Please, avoid using var. Assume the reader wants to know the type without seeing the return type of the method
- [ ] Surround exception-inducing code with `try-catch` clause, otherwise opening your project will lead to lots of problems highlighted in IDE (I used VS Code)
- [ ] Describe the dependency injection mechanisms you've used. I suspect you didn't have as many warnings in your IDE as I've described in comments

**Comment by 0202-D** [comment]
My level does not allow me to understand the code at the moment. I seriously tried it.

File: src/main/java/com/codereview/ipcounter/utils/Constants.java [hunk 26]
```diff
@@ -0,0 +1,9 @@
+package com.codereview.ipcounter.utils;
+
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+public class Constants {
+  public static final long NUMBER_OF_IP_ADDRESSES = 256L * 256 * 256 * 256;
```
Comment (KupriyanovAnton) [inline src/main/java/com/codereview/ipcounter/utils/Constants.java:7]: Аннотация UtilityClass автоматически помечает поля как static

File: src/main/java/com/codereview/ipcounter/utils/ParserUtils.java [hunk 27]
```diff
@@ -0,0 +1,16 @@
+package com.codereview.ipcounter.utils;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+public class ParserUtils {
+  public static long toLongValue(String ipString) throws UnknownHostException {
```
Comment (KupriyanovAnton) [inline src/main/java/com/codereview/ipcounter/utils/ParserUtils.java:9]: Поля тоже помечает как static
"All methods, inner classes, and fields in the class are marked static"

File: src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java (hunk 24 above)
Comment (MadPigeon) [reply src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java:13]: `1 << 27` means 1 * 2^27
Add a comment with this value (134217728). Describe why did you pick that value exactly

File: README.md (hunk 10 above)
Comment (AndryLat) [reply README.md:24]: Where did you try to call this command? And what was OS used? A lot of conditions. For example, for mac you should give execution rights, but macos user should know it:) readme must not be for begginers at all.

File: src/main/java/com/codereview/ipcounter/IpCounterApplication.java [hunk 28]
```diff
@@ -0,0 +1,13 @@
+package com.codereview.ipcounter;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class IpCounterApplication {
+
+  public static void main(String[] args) {
+    SpringApplication.run(IpCounterApplication.class, args);
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/IpCounterApplication.java:10]: In general, tab as 4 white spaces is preferred, but it is not big problem./

File: src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java (hunk 20 above)
Comment (AndryLat) [reply src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java:15]: It is entry point of application. In general, using Spring Boot for this application - it is huge overhead, but not big problem.

File: src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java (hunk 19 above)
Comment (AndryLat) [reply src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java:17]: All comments like this look like spam. I don't understand where did you take these warnings. My suggestion that you used IDE which doesn't understand Lombok + Spring autowiring. For example, here Spring will init this variable and will fail if it can't. What is the problem?

File: src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+import com.codereview.ipcounter.service.ArgumentProcessor;
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.boot.ApplicationArguments;
+import org.springframework.boot.context.event.ApplicationReadyEvent;
+import org.springframework.context.event.EventListener;
+import org.springframework.stereotype.Component;
+
+@Slf4j
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java:12]: You use logger but nothing is logged here! It is wrong because exactly here you can log a lot of useful information.

File: src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java [hunk 30]
```diff
@@ -0,0 +1,25 @@
...
+@RequiredArgsConstructor
+public class IpCounterEventHandler {
+
+  private final ApplicationArguments applicationArguments;
+  private final List<ArgumentProcessor> argumentProcessors;
+
+  @EventListener
+  public void onApplicationReadyEvent(ApplicationReadyEvent applicationReadyEvent) {
+    argumentProcessors.forEach(
+        argumentProcessor -> argumentProcessor.process(applicationArguments));
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java:23]: Can be done as 
image
Only proposition:)

File: src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java [hunk 31]
```diff
@@ -0,0 +1,25 @@
...
+  }
+}
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java:25]: Git doesn't like absence of empty line at the end of file.  Take into account and don't repeat. https://stackoverflow.com/questions/5813311/whats-the-significance-of-the-no-newline-at-end-of-file-log

File: src/main/java/com/codereview/ipcounter/service/ArgumentProcessor.java (hunk 14 above)
Comment (AndryLat) [reply src/main/java/com/codereview/ipcounter/service/ArgumentProcessor.java:6]: All functionality better to call via interface. it is right way, especially in Spring.

Comment (AndryLat) [reply src/main/java/com/codereview/ipcounter/service/ArgumentProcessor.java:6]: About name: it is not processor, it is parcer.

File: src/main/java/com/codereview/ipcounter/service/FileDataProcessor.java [hunk 32]
```diff
@@ -0,0 +1,6 @@
+package com.codereview.ipcounter.service;
+
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/service/FileDataProcessor.java:2]: use autoformatting everywhere. It is good practice:)

File: src/main/java/com/codereview/ipcounter/service/IpCounter.java (hunk 12 above)
Comment (AndryLat) [reply src/main/java/com/codereview/ipcounter/service/IpCounter.java:3]: I think generic is not needed at all.

File: src/main/java/com/codereview/ipcounter/service/IpCounter.java [hunk 33]
```diff
@@ -0,0 +1,10 @@
...
+
+  void countIp(T ipView);
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/service/IpCounter.java:5]: What does "ipView" mean? Not clear at all.

File: src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java [hunk 34]
```diff
@@ -0,0 +1,38 @@
...
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class FileIpDataProcessor implements FileDataProcessor {
+  private final IpCounter<String> ipCounter;
+
+  @Override
+  public void processFile(String filePath) {
+    try (Stream<String> ipLines = Files.lines(Paths.get(filePath))) {
+      var linesProcessed = new AtomicInteger();
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java:25]: Do we really need AtomicInteger?

File: src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java [hunk 35]
```diff
@@ -0,0 +1,38 @@
...
+      ipLines.takeWhile(line -> ipCounter.getCount() < NUMBER_OF_IP_ADDRESSES)
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java:26]: AS for me this condition is not clear. Some kind of optimisation? if yes it will break logic with showing "lines processed".

File: src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java [hunk 36]
```diff
@@ -0,0 +1,38 @@
...
+          .forEach(ipView -> {
+            ipCounter.countIp(ipView);
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java:28]: If I send to you only this line - you will never understand what's happened there. Name of variable should be improved (ipAddressString), countIp can be improved - "add", "ipCounter"
 -> uniqueIpCounter

File: src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java (hunk 22 above)
Comment (AndryLat) [reply src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java:31]: log will be init by Lombok.

Comment (AndryLat) [reply src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java:31]: result of app will be written in the log. Is it ok?

File: src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java [hunk 37]
```diff
@@ -0,0 +1,38 @@
...
+            linesProcessed.incrementAndGet();
+          });
+      log.info("Total lines processed: {} and unique ip count: {}", linesProcessed,
+          ipCounter.getCount());
+    } catch (IOException e) {
+      log.error("An error occurred during try to read file {}", filePath, e);
+      throw new IllegalStateException(e);
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/service/impl/FileIpDataProcessor.java:35]: Who\what will catch this exception?

File: src/main/java/com/codereview/ipcounter/service/impl/FileNameArgumentProcessor.java [hunk 38]
```diff
@@ -0,0 +1,31 @@
...
+import org.springframework.boot.ApplicationArguments;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class FileNameArgumentProcessor implements ArgumentProcessor {
+
+  private final FileDataProcessor fileDataProcessor;
+
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/service/impl/FileNameArgumentProcessor.java:19]: In general, there is strong code cohesion. It is very bad. Main logic should be in Handler where
first step: check args
second: open file
third: calculate
forth: show result.
And each part should be independent.

File: src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java (hunk 24 above)
Comment (AndryLat) [reply src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java:13]: It is copy\paste solution from Internet. Bad approach. in any case Vsevolod is right and if you use some not clear code in your solution, you should describe it somehow.

Comment (AndryLat) [reply src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java:13]: Also what does "seen" mean?

File: src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java (hunk 23 above)
Comment (AndryLat) [reply src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java:23]: Awful. You skip this ip and go next. Is it right logic?

File: src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java [hunk 39]
```diff
@@ -0,0 +1,45 @@
...
+public class UniqueIpCounter implements IpCounter<String> {
+
+  private final int[] seen = new int[1 << 27];
+
+  private long count;
+
+  @Override
+  public void countIp(String ipView) {
+    try {
+      var longIpView = ParserUtils.toLongValue(ipView);
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java:20]: ParserUtils.toLongValue(ipView); - say nothing! If I send you this part of code will you understand what is happend here? And if I send "IpAddressUtils.parseToLong(ipAddressStringValue)"?

File: src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java (hunk 25 above)
Comment (AndryLat) [reply src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java:37]: Name is not understandable. Logic of method - too.

File: src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java [hunk 40]
```diff
@@ -0,0 +1,45 @@
...
+    return count;
+  }
+
+  @Override
+  public void resetCounter() {
+    count = 0;
+  }
+
+  private void registerIfNewUnique(long longValue) {
+    int index = (int) (longValue >> 5);
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java:38]: Why 5? Why not 6?

Comment (AndryLat) [reply src/main/java/com/codereview/ipcounter/service/impl/UniqueIpCounter.java:38]: Magic numbers should be moved to constants and name should describe meaning of value.

File: src/main/java/com/codereview/ipcounter/utils/ParserUtils.java [hunk 41]
```diff
@@ -0,0 +1,16 @@
...
+    long result = 0;
+    for (byte b : InetAddress.getByName(ipString).getAddress()) {
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/utils/ParserUtils.java:11]: what does "b" mean? 
Parsing value via InetAddress is slow.

File: src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java [hunk 42]
```diff
@@ -0,0 +1,112 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.io.TempDir;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class FileIpDataProcessorUnitTest {
+
+  private static final String TEST_TXT = "test.txt";
```
Comment (AndryLat) [inline src/test/java/com/codereview/ipcounter/service/FileIpDataProcessorUnitTest.java:30]: name doesn't describe what is saved here.

File: src/test/java/com/codereview/ipcounter/utils/ParserUtilsUnitTest.java [hunk 43]
```diff
@@ -0,0 +1,23 @@
...
+
+  @SneakyThrows
+  @ParameterizedTest
```
Comment (AndryLat) [inline src/test/java/com/codereview/ipcounter/utils/ParserUtilsUnitTest.java:13]: Parameterized tests here should not be used. It is not useful. All test data groups can be checked in separate methods.

File: src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java [hunk 44]
```diff
@@ -0,0 +1,25 @@
+package com.codereview.ipcounter.handler;
```
Comment (AndryLat) [inline src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java:1]: what about package like "com.yourcodereview.leniz.task1..."?

File: README.md (hunk 10 above)
Comment (MadPigeon) [reply README.md:24]: Windows, PowerShell.

File: src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java (hunk 19 above)
Comment (MadPigeon) [reply src/main/java/com/codereview/ipcounter/handler/IpCounterEventHandler.java:17]: Thank you, I've installed Lombok and Spring autowiring extensions for VS code

File: src/main/java/com/codereview/ipcounter/service/FileDataProcessor.java (hunk 32 above)
Comment (MadPigeon) [reply src/main/java/com/codereview/ipcounter/service/FileDataProcessor.java:2]: What makes you understand autoformatting isn't used?
